org.eclipse.xtext/src/org/eclipse/xtext/resource/impl/DefaultResourceDescriptionManager.java

```diff
@@ -113,44 +117,62 @@ protected void addExportedNames(Set<QualifiedName> names, IResourceDescription r
-        for (IResourceDescription.Delta delta : deltas) {
-			if (hasChanges(delta, candidate)) {
-				// not a java resource - delta's resource should be contained in a visible container
-				// as long as we did not delete the resource
-				URI uri = delta.getUri();
-				if ((uri.isPlatform() || uri.isArchive()) && delta.getNew() != null) { 
-					if (containers == null)
-						containers = containerManager.getVisibleContainers(candidate, context);
-					boolean descriptionIsContained = false;
-					for(int i = 0; i < containers.size() && !descriptionIsContained; i++) {
-						descriptionIsContained = containers.get(i).hasResourceDescription(uri);
-					}
-					if (!descriptionIsContained)
-						return false;
-				}
-				if (isAffected(importedNames, delta.getNew()) || isAffected(importedNames, delta.getOld())) {
+		Set<URI> outgoingReferences = descriptionUtils.collectOutgoingReferences(candidate);
+		if (!outgoingReferences.isEmpty()) {
+			for (IResourceDescription.Delta delta : deltas)
+				if (hasChanges(delta, candidate) && outgoingReferences.contains(delta.getUri()))
 					return true;
+		}
+		// this is a tradeoff - we could either check whether a given delta uri is contained
+		// in a reachable container and check for intersecting names afterwards, or we can do
+		// the other way round
+		// unfortunately there is no way to decide reliably which algorithm scales better
+		// note that this method is called for each description so we have something like a
+		// number of deltas x number of resources which is not really nice
+		List<IContainer> containers = null;
+		Collection<QualifiedName> importedNames = getImportedNames(candidate);
+		if (!importedNames.isEmpty()) {
+			for (IResourceDescription.Delta delta : deltas) {
+				if (hasChanges(delta, candidate)) {
+					// not a java resource - delta's resource should be contained in a visible container
+					// as long as we did not delete the resource
+					URI uri = delta.getUri();
+					if ((uri.isPlatform() || uri.isArchive()) && delta.getNew() != null) {
+						if (containers == null)
+							containers = containerManager.getVisibleContainers(candidate, context);
+						boolean descriptionIsContained = false;
+						for (int i = 0; i < containers.size() && !descriptionIsContained; i++) {
+							descriptionIsContained = containers.get(i).hasResourceDescription(uri);
+						}
+						if (!descriptionIsContained) {
+							return isAffectedDueToExtension(deltas, candidate, context);
+						}
+					}
+					if (isAffected(importedNames, delta.getNew()) || isAffected(importedNames, delta.getOld())) {
+						return true;
+					}
 				}
 			}
-        }
-        return false;
-    }
+		}
+		return isAffectedDueToExtension(deltas, candidate, context);
+	}
+
+	/**
+	 * Query all registered extensions.
+	 * 
+	 * @since 2.22
+	 */
+	protected boolean isAffectedDueToExtension(Collection<Delta> deltas, IResourceDescription candidate,
```

Comment:
I think plural form `isAffectedDueToExtensions` could be a better choice here, since this method operates on a list of `IsAffectedExtension` objects.